### DISCUSSION TITLE
Enhance `mkdocs serve` behavior?

### DISCUSSION THREAD

**Discussion by fralau** [mkdocs/mkdocs#3494] [Ideas]
Unsurprisingly, `mkdocs serve` is one of the most frequently used commands with MkDocs.

There are are two improvements I would like to see:

1. Start the default browser. That seems a reasonable approach, as done by Jupyter for example.
2. If the default port is taken, don't fail; just take the next one.

I haven't made a survey, but I suppose that a good 80% of the users would be happy with that evolution, out of the box.

There might always be a number who do not like it, for legitimate reasons (it would be good to list those objections, to understand how to answer them). To avoid taking risks I would be happy with the following approach:

1. Initially, make the new behavior optional (opt-in), with a parameter in the config file
2. Then see how it goes, and if it seems a good idea, make it standard.
3. But if there the slightest doubt, allow an opt-out in the config file.

What would you think?

**Comment by oprypin** [comment]
These are good ideas. You could open an issue for each of them.
Regarding ports it'll need some discussion of details to not make it the port totally unpredictable

**Reply by oprypin** [reply]
I'll implement these

**Reply by fralau** [reply]
Opened:
- #3495 for starting the default browser
- #3496 for port switching

**Reply by oprypin** [reply]
Opened pull request:
* #3497

**Comment by pawamoy** [comment]
Not a fan of automatically opening the URL in the browser. First because I use different browser profiles and it probably won't use the right one, and second because even if it got the right profile, I don't want it to reopen the same tab each time I interrupt and re-run `mkdocs serve`. If it can detect that the URL is already open in a tab, then that would be best. Anyway, I'm fine with either opt-in or opt-out :slightly_smiling_face:

**Comment by squidfunk** [comment]
I'm pretty sure this could be implemented in a plugin using the [`on_serve`](https://www.mkdocs.org/dev-guide/plugins/#on_serve) hook which is executed once the initial build is finished, and allows you to register additional logic when file changes are detected.

**Reply by fralau** [reply]
This is an intriging idea. It could work, at least for starting the browser. 🤔

**Reply by squidfunk** [reply]
Jup, the "choose another port" behavior might not be implementable, because `on_startup` doesn't receive the `MkDocsConfig` object, which is the only hook that one could use to dynamically change the `dev_addr` setting. Maybe this might be added to core, i.e., providing it as a third option to the hook.

https://github.com/mkdocs/mkdocs/blob/646987da4502becf346bfaabb6ba40934307399b/mkdocs/commands/serve.py#L56-L69

**Comment by redtide** [comment]
> 1. Start the default browser. That seems a reasonable approach, as done by Jupyter for example.

This sounds strange to me, why a page server should open a client? It seems to me so straightforward as is: start serving pages, click on the link that opens the browser (it's a user choice to do it or not). When restarting there is already livereload, which I would like to make optional but this is another thing, otherwise just press F5.

> 2. If the default port is taken, don't fail; just take the next one.

Until now I haven't miss this neither when using Jekyll, but I don't see why not having it as option (I mean, OK to me if will be part of mkdocs, but I would be against to have it enabled by default).
It could be implemented as a list of regexes maybe? So you can also choose the way it selects the port number.
